src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java

```diff
@@ -146,4 +154,89 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
             return Mth.floor(proportion * 14.0F) + (itemsFound > 0 ? 1 : 0);
         }
     }
+
+    /**
+     * Try to move items from one inventory to another.
+     *
+     * @param from          source inventory
+     * @param to            destination inventory
+     * @param filter        a filter to select which items to move; if the filter returns {@code false} for a stack, it won't be moved
+     * @param maxAmount     maximum total amount of items to move
+     * @param stackInTarget {@code true} to try to stack with existing items in the target inventory,
+     *                      {@code false} to insert in the first slot that fits
+     * @return the total amount of moved items
+     */
+    public static int moveItems(
+            @Nullable IItemHandler from,
+            @Nullable IItemHandler to,
+            Predicate<ItemStack> filter,
+            int maxAmount,
+            boolean stackInTarget) {
+        Objects.requireNonNull(filter, "filter");
+        if (from == null || to == null || maxAmount <= 0) {
+            return 0;
+        }
+
+        int totalMoved = 0;
+
+        int fromSlots = from.getSlots();
+        for (int i = 0; i < fromSlots; ++i) {
+            // Check filter
+            if (!filter.test(from.getStackInSlot(i))) {
+                continue;
+            }
+
+            // Repeated extraction because extractItem limits to max stack size
+            while (true) {
+                // Simulate extraction
+                var available = from.extractItem(i, Integer.MAX_VALUE, true);
+                if (available.isEmpty()) {
+                    break;
+                }
+                int availableCount = available.getCount();
+
+                // Simulate insertion
+                var simulationLeftover = stackInTarget ? insertItemStacked(to, available, true) : insertItem(to, available, true);
+                int canFit = availableCount - simulationLeftover.getCount();
```

Comment:
From experience, this number will be extremely unreliable. Simulating inserting items into multiple slots is a bad idea, there are many targets where the slots are not independent. You need a copy of insertItemStacked that returns the number it can insert into the first target it can insert into. Then execute that insert and repeat until maxItems or source empty.

One simple example would be the Ender IO Alloy Smelter. It has 3 input slots, and if you use insertItemStacked with Iron Ingots, it will report that it can insert 192 items, i.e. a full stack in each slot. But, as soon as there's iron in one slot, it will lock in a recipe and the other two slots won't accept any iron.

Here's some proven code: https://github.com/SleepyTrousers/EnderIO-1.5-1.12/blob/master/enderio-base/src/main/java/crazypants/enderio/base/capability/ItemTools.java#L370 (Line 170 has the moveItems() equivalent, although it's a bit less readable as it has a couple extra features). And even with that we had spilled items on occasion.

Reply:
Yes you are correct. Choosing how to handle these cases is the difficult aspect of this PR, and it highlights how difficult `IItemHandler` is to compose with higher-level abstractions. (`insertItemStacked` is kinda broken)

Reply:
It works sufficiently well if you stop after the first successful simulated operation. Because that should be repeatable and work for a non-simulated identical operation that is executed in the same context.

There still are one logic issue and one performance issue:

- The actual operation must be identical, i.e. the itemstack size needs to be the same as in the simulation. A smaller size may lead to different results (it's rare, but only being able to hold, e.g., integer multiples of recipe amounts is legal). So, using the simulation to determine the maximum insertable size and then min() it with other constraints doesn't work.
- Running the simulation over the target slot by slot to find one where it can insert, and then doing the same thing for the actual insert wastes CPU (remember that Giant Chests debacle?). The slot number of the slot where the simulation succeeded should be reused.

Note that the code I linked is guilty of both. Those optimisations are in the conduit code, but that code isn't readable outside that context. I never had a need to put them into the machine pull/push code, as they're not that critical there.

---

Warstory Mode: Giant Chests (I think that was what the mod was called) got included in a popular modpack (I think it was one of the skyblock ones) and got featured by basically every YouTuber playing that pack. However, it quickly turned out that there was not a single item transport solution that wouldn't lag the server badly trying to insert into its flat inventory with thousands of slots. Server operators quickly banned that mod and (some) item transport mods slowly implemented slot scanning limits. But those require the pipe to remember where it stopped scanning for insertable slots in the previous tick, making the whole logic even more complicated. If you ever wondered why an Item Conduit keeps adding to the same stack, even if you manually empty a slot before that one, something a Hopper does not, this is why.